core/src/main/java/hudson/lifecycle/Lifecycle.java

```diff
@@ -52,6 +52,19 @@
 public abstract class Lifecycle implements ExtensionPoint {
     private static Lifecycle INSTANCE = null;
 
+    public Lifecycle() {
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            Jenkins jenkins = Jenkins.getInstanceOrNull();
+            if (jenkins != null) {
+                try {
+                    jenkins.cleanUp();
+                } catch (Throwable e) {
+                    LOGGER.log(Level.SEVERE, "Failed to clean up. Shutdown will continue.", e);
+                }
+            }
+        }));
+    }
```

Comment:
This PR is downstream of #6228, which calls `Jenkins.get().getLifecycle()` at the completion of successful startup. Without that, this constructor wouldn't be called early enough to matter.